File: reference/ts-solver/tests/core/heuristic.test.ts
import { describe, expect, it } from 'bun:test';
import { buildAspectData } from '../../app/src/data/aspects';
import { makeInventory, DEFAULT_THRESHOLD } from '../../app/src/core/inventory';
import { createBoard, setState } from '../../app/src/core/board';
import { remainderHeuristic } from '../../app/src/core/heuristic';

const data = buildAspectData();

describe('remainderHeuristic (admissible, spec §5.2)', () => {
  it('is (0,0) when all anchors are already connected', () => {
    const b = createBoard(2);
    setState(b, { q: 0, r: 0 }, { kind: 'ANCHOR', aspect: 'air' });
    setState(b, { q: 1, r: 0 }, { kind: 'PLACED', aspect: 'void', locked: false });
    setState(b, { q: 2, r: 0 }, { kind: 'ANCHOR', aspect: 'entropy' });
    const inv = makeInventory([['air', 100], ['entropy', 100]], DEFAULT_THRESHOLD);
    const h = remainderHeuristic(data, b, inv);
    expect(h.cells).toBe(0);
    expect(h.scarcity).toBe(0);
  });

  it('needs >=1 inner cell for two anchors at distance 2 (hCells >= 1)', () => {
    const b = createBoard(2);
    setState(b, { q: 0, r: 0 }, { kind: 'ANCHOR', aspect: 'air' });
    setState(b, { q: 2, r: 0 }, { kind: 'ANCHOR', aspect: 'entropy' });
    const inv = makeInventory([['air', 100]], DEFAULT_THRESHOLD);
    const h = remainderHeuristic(data, b, inv);
    expect(h.cells).toBeGreaterThanOrEqual(1);
  });

  it('routes around dead hexes (no path => Infinity scarcity)', () => {
    const b = createBoard(2);
    setState(b, { q: 0, r: 0 }, { kind: 'ANCHOR', aspect: 'air' });
    setState(b, { q: 2, r: 0 }, { kind: 'ANCHOR', aspect: 'entropy' });
    // wall off (2,0) entirely with DEAD neighbors so it cannot be reached
    for (const n of [{ q: 1, r: 0 }, { q: 2, r: -1 }, { q: 1, r: 1 }]) setState(b, n, { kind: 'DEAD' });
    const inv = makeInventory([['air', 100]], DEFAULT_THRESHOLD);
    const h = remainderHeuristic(data, b, inv);
    expect(h.scarcity).toBe(Number.POSITIVE_INFINITY);
  });

  it('does not treat a locked-only island as a terminal (h stays finite/small)', () => {
    const b = createBoard(3);
    setState(b, { q: 0, r: 0 }, { kind: 'ANCHOR', aspect: 'air' });
    setState(b, { q: 1, r: 0 }, { kind: 'ANCHOR', aspect: 'fire' }); // air-fire? not an edge -> but heuristic ignores labels
    setState(b, { q: -3, r: 3 }, { kind: 'PLACED', aspect: 'water', locked: true }); // far island
    const inv = makeInventory([['air', 100]], DEFAULT_THRESHOLD);
    const h = remainderHeuristic(data, b, inv);
    // island must NOT force a long connection; with anchors adjacent, hCells should be 0
    expect(h.cells).toBe(0);
  });
});

File: src/test/kotlin/io/github/muindor/tcresearchsolver/solver/HeuristicTest.kt
package io.github.muindor.tcresearchsolver.solver

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class HeuristicTest {

    private val data = buildAspectData()

    @Test
    fun `is (0,0) when all anchors are already connected`() {
        val b = createBoard(2)
        setState(b, Hex(0, 0), CellState.Anchor("air"))
        setState(b, Hex(1, 0), CellState.Placed("void", false))
        setState(b, Hex(2, 0), CellState.Anchor("entropy"))
        val inv = makeInventory(listOf("air" to 100, "entropy" to 100), DEFAULT_THRESHOLD)
        val h = remainderHeuristic(data, b, inv)
        assertEquals(0.0, h.cells, "cells should be 0")
        assertEquals(0.0, h.scarcity, "scarcity should be 0")
    }

    @Test
    fun `needs at least 1 inner cell for two anchors at distance 2`() {
        val b = createBoard(2)
        setState(b, Hex(0, 0), CellState.Anchor("air"))
        setState(b, Hex(2, 0), CellState.Anchor("entropy"))
        val inv = makeInventory(listOf("air" to 100), DEFAULT_THRESHOLD)
        val h = remainderHeuristic(data, b, inv)
        assertTrue(h.cells >= 1.0, "cells should be >= 1 but was ${h.cells}")
    }

    @Test
    fun `routes around dead hexes - no path gives Infinity scarcity`() {
        val b = createBoard(2)
        setState(b, Hex(0, 0), CellState.Anchor("air"))
        setState(b, Hex(2, 0), CellState.Anchor("entropy"))
        // wall off (2,0) entirely with DEAD neighbors so it cannot be reached
        for (n in listOf(Hex(1, 0), Hex(2, -1), Hex(1, 1))) setState(b, n, CellState.Dead)
        val inv = makeInventory(listOf("air" to 100), DEFAULT_THRESHOLD)
        val h = remainderHeuristic(data, b, inv)
        assertEquals(Double.POSITIVE_INFINITY, h.scarcity, "scarcity should be +Infinity when no path exists")
    }

    @Test
    fun `does not treat a locked-only island as a terminal (h stays finite or small)`() {
        val b = createBoard(3)
        setState(b, Hex(0, 0), CellState.Anchor("air"))
        setState(b, Hex(1, 0), CellState.Anchor("fire"))
        setState(b, Hex(-3, 3), CellState.Placed("water", true)) // far island, locked but no anchor
        val inv = makeInventory(listOf("air" to 100), DEFAULT_THRESHOLD)
        val h = remainderHeuristic(data, b, inv)
        // island must NOT force a long connection; with anchors adjacent, hCells should be 0
        assertEquals(0.0, h.cells, "cells should be 0 when anchors are adjacent")
    }
}
